common/client/multi_node_test.go

```diff
@@ -0,0 +1,629 @@
+	types.Receipt[Hashable, Hashable], Hashable, types.Head[Hashable]] {
+	return newMockRPC[types.ID, *utils.Big, Hashable, Hashable, any, Hashable, any, any,
+		types.Receipt[Hashable, Hashable], Hashable, types.Head[Hashable]](t)
+}
+
+func newHealthyNode(t *testing.T, chainID types.ID) *mockNode[types.ID, types.Head[Hashable], multiNodeRPCClient] {
+	return newNodeWithState(t, chainID, nodeStateAlive)
+}
+
+func newNodeWithState(t *testing.T, chainID types.ID, state nodeState) *mockNode[types.ID, types.Head[Hashable], multiNodeRPCClient] {
+	node := newMockNode[types.ID, types.Head[Hashable], multiNodeRPCClient](t)
+	node.On("ConfiguredChainID").Return(chainID).Once()
+	node.On("Start", mock.Anything).Return(nil).Once()
+	node.On("Close").Return(nil).Once()
+	node.On("State").Return(state).Maybe()
+	node.On("String").Return(fmt.Sprintf("healthy_node_%d", rand.Int())).Maybe()
+	return node
+}
+func TestMultiNode_Dial(t *testing.T) {
+	t.Parallel()
+
+	newMockNode := newMockNode[types.ID, types.Head[Hashable], multiNodeRPCClient]
+	newMockSendOnlyNode := newMockSendOnlyNode[types.ID, multiNodeRPCClient]
+
+	t.Run("Fails without nodes", func(t *testing.T) {
+		t.Parallel()
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       types.RandomID(),
+		})
+		err := mn.Dial(tests.Context(t))
+		assert.EqualError(t, err, fmt.Sprintf("no available nodes for chain %s", mn.chainID.String()))
+	})
+	t.Run("Fails with wrong node's chainID", func(t *testing.T) {
+		t.Parallel()
+		node := newMockNode(t)
+		multiNodeChainID := types.NewIDFromInt(10)
+		nodeChainID := types.NewIDFromInt(11)
+		node.On("ConfiguredChainID").Return(nodeChainID).Twice()
+		const nodeName = "nodeName"
+		node.On("String").Return(nodeName).Once()
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       multiNodeChainID,
+			nodes:         []Node[types.ID, types.Head[Hashable], multiNodeRPCClient]{node},
+		})
+		err := mn.Dial(tests.Context(t))
+		assert.EqualError(t, err, fmt.Sprintf("node %s has configured chain ID %s which does not match multinode configured chain ID of %s", nodeName, nodeChainID, mn.chainID))
+	})
+	t.Run("Fails if node fails", func(t *testing.T) {
+		t.Parallel()
+		node := newMockNode(t)
+		chainID := types.RandomID()
+		node.On("ConfiguredChainID").Return(chainID).Once()
+		expectedError := errors.New("failed to start node")
+		node.On("Start", mock.Anything).Return(expectedError).Once()
+		mn := newTestMultiNode(t, multiNodeOpts{
+			selectionMode: NodeSelectionModeRoundRobin,
+			chainID:       chainID,
+			nodes:         []Node[types.ID, types.Head[Hashable], multiNodeRPCClient]{node},
+		})
+		err := mn.Dial(tests.Context(t))
+		assert.EqualError(t, err, expectedError.Error())
+	})
+
+	t.Run("Closes started nodes on failure", func(t *testing.T) {
```

Comment:
I think this name is a bit misleading. During `Dial` if a node tries to get started and it fails for any reason besides a wrong chainID, it will simply return an error. The node was never properly started to be closed. Perhaps something like this would be more accurate:
 ```suggestion
	t.Run("Fails if at least one node fails", func(t *testing.T) {
```

Reply:
We test multinode failure on node failure above. Here the idea was to ensure that we properly release resources of node1, that was started if node2 fails. `newHealthyNode` adds expectation of `Close` call

Reply:
I see. `newHealthyNode` seemed like a method that creates a mock node with the right configs. The expectation of `Start` and `Close` was a bit more difficult to figure out.